src/compute-client/src/controller/replicated.rs

```diff
@@ -23,189 +23,111 @@
-        loop {
-            state = match state {
-                ReplicaTaskState::Reconnect => self.step_reconnect().await,
-                ReplicaTaskState::Drain => self.step_drain().await,
-                ReplicaTaskState::Pump => self.step_pump().await,
-                ReplicaTaskState::Done => break,
-            }
-        }
-        debug!("task for replica {} exiting", self.replica_id);
+    let replica_id = config.replica_id;
+    info!("starting replica task for {replica_id}");
+    match run_replica_core(config).await {
+        Ok(()) => info!("gracefully stopping replica task for {replica_id}"),
+        Err(e) => warn!("replica task for {replica_id} failed: {e}"),
     }
+}
 
-    async fn step_reconnect(&mut self) -> ReplicaTaskState {
-        if let Some(last_successful_connection) = self.last_successful_connection {
-            // If we were previously up for long enough (60 seconds chosen
-            // arbitrarily), we consider the previous connection to have been
-            // successful and reset the backoff.
-            if Instant::now() - last_successful_connection > Duration::from_secs(60) {
-                self.backoff = INITIAL_CONNECTION_BACKOFF;
-            }
-        }
-        time::sleep(self.backoff).await;
-        // 60 seconds is arbitrarily chosen as a maximum plausible backoff. If
-        // Timely processes aren't realizing their buddies are down within that
-        // time, something is seriously hosed with the network anyway and its
-        // unlikely things will work.
-        self.backoff = (self.backoff * 2).min(Duration::from_secs(60));
-        match self.client.reconnect().await {
-            Ok(()) => {
-                self.last_successful_connection = Some(Instant::now());
-                ReplicaTaskState::Drain
-            }
-            Err(e) => {
-                info!(
-                    "Reconnecting to replica {} failed (retrying in {:?}): {}",
-                    self.replica_id, self.backoff, e
-                );
-                ReplicaTaskState::Reconnect
+async fn run_replica_core<T>(
+    ReplicaTaskConfig {
+        replica_id,
+        addrs,
+        mut command_rx,
+        response_tx,
+    }: ReplicaTaskConfig<T>,
+) -> Result<(), anyhow::Error>
+where
+    T: Timestamp + Copy,
+    ComputeGrpcClient: ComputeClient<T>,
+{
+    let mut client = Retry::default()
+        .clamp_backoff(Duration::from_secs(32))
+        .retry_async(|state| {
+            let addrs = addrs.clone();
+            async move {
+                match ComputeGrpcClient::connect_partitioned(addrs).await {
+                    Ok(client) => Ok(client),
+                    Err(e) => {
+                        warn!(
+                            "error connecting to replica {replica_id}, retrying in {:?}: {e}",
+                            state.next_backoff.unwrap()
+                        );
+                        Err(e)
+                    }
+                }
```

Comment:
I think this prevents replicas from being dropped if they never succeed in connecting to any peer. Even if dropped, they don't see it because they don't try to receive on their command channel.

Reply:
I filed https://github.com/MaterializeInc/materialize/issues/13668 about this.